cdap-common/bin/functions.sh

```diff
@@ -493,11 +491,34 @@ cdap_set_spark() {
         return 1
       fi
       SPARK_HOME=$(echo -e "${SPARK_VAR_OUT}" | grep ^SPARK_HOME= | cut -d= -f2)
-      export SPARK_HOME
-      return 0
     fi
-    return 1
   fi
+
+  export SPARK_HOME
+
+  # Find environment variables setup via spark-env.sh
+  SPARK_ENV_SH=${SPARK_HOME}/conf/spark-env.sh
+  if [[ -f ${SPARK_ENV_SH} ]]; then
+    if [[ -z ${SPARK_ENV_PATTERN} ]]; then
+      # By default, only get env from spark-env.sh that starts with PY or SPARK
+      SPARK_ENV_PATTERN="^(PY|SPARK)"
+    fi
+    SPARK_ENV=$(source ${SPARK_ENV_SH}; env | grep -E "${SPARK_ENV_PATTERN}")
+    __ret=$?
+    if [[ ${__ret} -eq 0 ]]; then
+      local readonly __env_out=$(mktemp)
+      echo "${SPARK_ENV}" | while read -r line; do
+        # Prefix the env variable with _SPARK_ to avoid conflicts
+        echo "export _SPARK_${line%%=*}=${line#*=}" >> ${__env_out}
+      done
+      source ${__env_out}
+      rm ${__env_out}
```

Comment:
To an extent, yes. They exist above, too, but we only ever write error messages to the file to display later. If some joker manages to replace that with a link to /etc/passwd, it won't *break* anything, per se. If someone manages to change it here, we're going to use it to execute code. If we keep everything in the shell, that doesn't happen.

It's probably easiest to simply export all of the matching variables from the grep into the current shell, and pass that through, rather than prefixing them. CDAP itself doesn't run Spark code, so the variables won't effect its behavior.